rust/protocol/src/sealed_sender.rs

```diff
@@ -879,8 +1031,133 @@ mod sealed_sender_v2 {
+        let sender_at_0 = compute_authentication_tag(
+            &sender_identity,
+            recipient_identity.identity_key(),
+            Direction::Sending,
+            &ephemeral_public_key,
+            &sender_c_0,
+        )?;
+
+        // The message recipient calculates the original random bytes and authenticates the result.
+        let recv_m = apply_agreement_xor(
+            &recipient_keypair,
+            &ephemeral_public_key,
+            Direction::Receiving,
+            &sender_c_0,
+        )?;
+        assert_eq!(recv_m.as_ref(), m.as_ref());
+
+        let recv_at_0 = compute_authentication_tag(
+            &recipient_identity,
+            sender_identity.identity_key(),
+            Direction::Receiving,
+            &ephemeral_public_key,
+            &sender_c_0,
+        )?;
+        assert_eq!(recv_at_0.as_ref(), sender_at_0.as_ref());
+
+        Ok(())
+    }
 }
 
+/// This method implements a single-key multi-recipient [KEM], defined in Manuel Barbosa's
+/// ["Randomness Reuse: Extensions and Improvements"].
+///
+/// [KEM]: https://en.wikipedia.org/wiki/Key_encapsulation
+/// ["Randomness Reuse: Extensions and Improvements"]: https://haslab.uminho.pt/mbb/files/reuse.pdf
+///
+/// ### Contrast with Single-Recipient KEM
+/// The KEM scheme implemented by this method uses the "Generic Construction" in `4.1` of that
+/// paper, instantiated with [ElGamal encryption]. Unlike other message types sent over the wire,
+/// the encoded message returned by this method does *not* correspond to a specific protobuf schema
+/// due to the complex web of cryptographic data dependencies in different segments of the resulting
+/// encoded message. See the documentation of our [single-recipient KEM] for a discussion of the
+/// design tradeoffs of using this particular multi-recipient KEM.
+///
+/// [ElGamal encryption]: https://en.wikipedia.org/wiki/ElGamal_encryption
+/// [single-recipient KEM]: sealed_sender_encrypt_from_usmc#contrast-with-multi-recipient-kem
+///
+/// ### Overview
+/// The KEM scheme described in `4.1` of ["Randomness Reuse: Extensions and Improvements"] and
+/// implemented in this method enables reusing a single sequence of random bytes across multiple
+/// messages with the same content. This reduces computation time for clients sending the same
+/// message to multiple recipients without compromising the message security. The high-level steps
+/// of this process are listed below:
+/// 1. Generate a series of random bytes.
+/// 2. Derive an ephemeral key pair from (1).
+/// 3. Encrypt (1) using a shared secret derived from the private ephemeral key (2) and the
+///    recipient's public identity key.
+/// 4. Add an authentication tag with for (3) using a secret derived from the sender's private
+///    identity key and the recipient's public identity key.
+/// 5. Use the keypair from (2) to symmetrically encrypt the underlying
+///    [UnidentifiedSenderMessageContent] via [AEAD encryption].
```

Comment:
Let's add that this only has to be done once per recipient, unlike (3) and (4).

Reply:
Thanks! Done.